pandas/core/internals.py

```diff
@@ -4686,72 +4704,59 @@ def form_blocks(arrays, names, axes):
         k = names[name_idx]
         v = arrays[name_idx]
 
-        if is_sparse(v):
-            sparse_items.append((i, k, v))
-        elif issubclass(v.dtype.type, np.floating):
-            float_items.append((i, k, v))
-        elif issubclass(v.dtype.type, np.complexfloating):
-            complex_items.append((i, k, v))
-        elif issubclass(v.dtype.type, np.datetime64):
-            if v.dtype != _NS_DTYPE:
-                v = conversion.ensure_datetime64ns(v)
-
-            if is_datetimetz(v):
-                datetime_tz_items.append((i, k, v))
-            else:
-                datetime_items.append((i, k, v))
-        elif is_datetimetz(v):
-            datetime_tz_items.append((i, k, v))
-        elif issubclass(v.dtype.type, np.integer):
-            int_items.append((i, k, v))
-        elif v.dtype == np.bool_:
-            bool_items.append((i, k, v))
-        elif is_categorical(v):
-            cat_items.append((i, k, v))
-        else:
-            object_items.append((i, k, v))
+        block_type = _get_block_type(v)
+
+        if block_type == 'datetime' and v.dtype != _NS_DTYPE:
```

Comment:
then don't do it

Reply:
The status quo does this, so it merits double-checking.  If we remove this line, then in the relevant case we will end up calling (inside simple_blockify) `v = v.astype(_NS_DTYPE)` which should be equivalent to this, but presumably less performant.  I'm OK with that, but whoever put this here might have had a reason.